ukvm/ukvm_hv_openbsd_x86_64.c

```diff
@@ -0,0 +1,254 @@
+    if (vrp == NULL)
+        err(1, "malloc vrp");
+
+    vrp->vrp_exit = malloc(sizeof(union vm_exit));
+    if (vrp == NULL)
+        err(1, "malloc vrp_exit");
+
+    vrp->vrp_vm_id = hvb->vcp_id;
+    vrp->vrp_vcpu_id = hvb->vcpu_id;
+    vrp->vrp_continue = 0;
+
+    if ((pw = getpwnam(VMD_USER)) == NULL)
+        err(1, "can't get _vmd user");
+    uid = pw->pw_uid;
+    gid = pw->pw_gid;
+
+    if (chroot(pw->pw_dir) == -1)
+        err(1, "chroot: %s", pw->pw_dir);
+
+    if (chdir("/") == -1)
+        err(1, "chdir(\"/\")");
+
+    if (setgroups(1, &gid) ||
+        setresgid(gid, gid, gid) ||
+        setresuid(uid, uid, uid))
+        err(1, "unable to revoke privs");
+
+    /*
+     * pledge in the vm processes:
+     * stdio - for malloc and basic I/O including events.
+     * vmm - for the vmm ioctls and operations.
+     */
+    if (pledge("stdio vmm", NULL) == -1)
+        err(errno, "pledge");
+
+    for (;;) {
+        vrp->vrp_irq = 0xFFFF;
+
+        if (ioctl(hvb->vmd_fd, VMM_IOC_RUN, vrp) < 0) {
+            if(errno == EIO) // guest halted with interrupts disabled, ie cli, hlt
+                return;
+
+            err(1, "ukvm_hv_vcpu_loop: vm / vcpu run ioctl failed");
+        }
+
+        /* If the VM is terminating, exit normally */
+        if (vrp->vrp_exit_reason == VM_EXIT_TERMINATED) {
+            return;
+        }
+
+        union vm_exit *vei = vrp->vrp_exit;
+
+        if (vrp->vrp_exit_reason != VM_EXIT_NONE) {
+            switch (vrp->vrp_exit_reason) {
+                case VMX_EXIT_INT_WINDOW:
+                case VMX_EXIT_CPUID:
+                case VMX_EXIT_EXTINT:
+                case SVM_VMEXIT_INTR:
+                case VMX_EXIT_EPT_VIOLATION:
+                case SVM_VMEXIT_NPF:
+                case SVM_VMEXIT_MSR:
+                case SVM_VMEXIT_CPUID:
+                    // nothing to be done here, as per vmd
+                    break;
+                case VMX_EXIT_IO:
+                case SVM_VMEXIT_IOIO:
+                    if (vei->vei.vei_dir != VEI_DIR_OUT
+                            || vei->vei.vei_size != 4)
+                        errx(1, "Invalid guest port access: port=0x%x", vei->vei.vei_port);
+                    if (vei->vei.vei_port < UKVM_HYPERCALL_PIO_BASE ||
+                            vei->vei.vei_port >= (UKVM_HYPERCALL_PIO_BASE + UKVM_HYPERCALL_MAX))
+                        errx(1, "Invalid guest port access: port=0x%x", vei->vei.vei_port);
+
+                    int nr = vei->vei.vei_port - UKVM_HYPERCALL_PIO_BASE;
+                    ukvm_hypercall_fn_t fn = ukvm_core_hypercalls[nr];
+                    if (fn == NULL)
+                        errx(1, "Invalid guest hypercall: num=%d", nr);
+
+                    ukvm_gpa_t gpa = vei->vei.vei_data;
+                    fn(hv, gpa);
+                    break;
+                case VMX_EXIT_HLT:
+                case SVM_VMEXIT_HLT:
+                    // stop on exit
```

Comment:
Is this the same as the `EIO` case further up? What's the difference?

Reply:
again following vmd [vcpu_exit#L1485-L1499](https://github.com/openbsd/src/blob/7e7a8c998653110b57ed8bc9fdb7cb231e91d0a8/usr.sbin/vmd/vm.c#L1485-L1499)

i might need to look further into these two, and ask some questions.

Reply:
base on [vmm.c : vmx_handle_halt](https://github.com/openbsd/src/blob/master/sys/arch/amd64/amd64/vmm.c#L4237-L4283) first the return value comments and second line [L4275-L4282](https://github.com/openbsd/src/blob/master/sys/arch/amd64/amd64/vmm.c#L4275-L4282) The EIO case is for when the guest halts with interrupts disabed, (the case for solo5) and the this case is when vmm has receive a VMX_EXIT_HLT exit reason but the guest has not halted with interrupts disabled, but i am not sure if this will ever occur.

based on this, we could remove these lines and if an occurrence pops up we can always add them back.

Reply:
Yes, let's do that.